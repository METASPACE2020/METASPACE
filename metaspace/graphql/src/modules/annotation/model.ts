import {
  Entity,
  PrimaryColumn,
  Column, ManyToOne, JoinColumn, OneToMany, JoinTable, Index, PrimaryGeneratedColumn,
} from 'typeorm';
import {MomentValueTransformer} from '../../utils/MomentValueTransformer';
import {Moment} from 'moment';

@Entity('coloc_job')
export class ColocJob {

  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', name: 'ds_id' })
  datasetId: string;

  @Column({ type: 'text', name: 'mol_db', nullable: true })
  molDb: string | null;

  @Column({ type: 'numeric', precision: 2, scale: 2 })
  fdr: number;

  @Column({ type: 'text' })
  algorithm: string;

  @Column({ type: 'timestamp without time zone', default: () => "(now() at time zone 'utc')",
    transformer: new MomentValueTransformer() })
  start: Moment;

  @Column({ type: 'timestamp without time zone', default: () => "(now() at time zone 'utc')",
    transformer: new MomentValueTransformer() })
  finish: Moment;

  @Column({ type: 'text', nullable: true })
  error: string | null;

  @Column({ type: 'int', array: true, name: 'sample_ion_ids' })
  sampleIonIds: number[];

  @OneToMany(type => ColocAnnotation, colocAnnotation => colocAnnotation.colocJob)
  @JoinTable({ name: 'dataset_project' })
  colocAnnotations: ColocAnnotation[];
}

@Entity('coloc_annotation')
export class ColocAnnotation {

  @PrimaryColumn({ type: 'uuid', name: 'coloc_job_id' })
  colocJobId: string;

  /** Index of the annotation's sfAdduct in the parent ColocJob's orderedMols array */
  @PrimaryColumn({ type: 'int', name: 'ion_id' })
  ionId: number;

  @Column({ type: 'int', array: true, name: 'coloc_ion_ids' })
  colocIonIds: number[];

  @Column({ type: 'float4', array: true, name: 'coloc_coeffs' })
  colocCoeffs: number[];

  @ManyToOne(type => ColocJob, {onDelete: 'CASCADE'})
  @JoinColumn({ name: 'coloc_job_id' })
  colocJob: ColocJob;
}

@Entity('ion')
// WIP: This index doesn't work. Postgres treats all nulls as unequal, meaning any row that contains at least one null
// will pass the uniqueness check, even if there is an identical row already.
@Index(['formula', 'chemMod', 'neutralLoss', 'adduct', 'charge'], {unique: true})
export class Ion {
  @PrimaryGeneratedColumn({ type: 'int' })
  id: number;

  /** All components of the ion in one string, with charge as a '+'/'-' suffix e.g. "C21H41O6P-HO+HO3P-H2O-CO2+Na+" */
  @Column({ type: 'text' })
  @Index()
  ion: string;

  /** The formula of the molecule prior to any modifications, e.g. "C21H41O6P".
   *  Internally this has been called "sumFormula" or "sf", but this term is incorrect.
   *  It should be referred to as "formula" or "molecular formula". */
  @Column({ type: 'text' })
  formula: string;

  @Column({ type: 'text', nullable: true, name: 'chem_mod' })
  chemMod: string;

  @Column({ type: 'text', nullable: true, name: 'neutral_loss' })
  neutralLoss: string;

  @Column({ type: 'text' })
  adduct: string;

  /** 1 or -1 */
  @Column({ type: 'smallint' })
  charge: number;
}
