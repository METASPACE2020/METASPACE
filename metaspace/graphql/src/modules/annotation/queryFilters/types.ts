import {ArgsFromBinding} from '../../../bindingTypes';
import {Query} from '../../../binding';
import {ESAnnotation} from '../../../../esConnector';

export type Args = ArgsFromBinding<Query['allAnnotations']>
  | ArgsFromBinding<Query['countAnnotations']>;

export type PostProcessFunc = (annotations: ESAnnotation[]) => ESAnnotation[];

export interface FilterResult {
  args: Args;
  postprocess?: PostProcessFunc;
}

export interface ESAnnotationWithColoc extends ESAnnotation {
  _cachedColocCoeff: number | null;
  _isColocReference: boolean;

  getColocalizationCoeff(_colocalizedWith: string, _colocalizationAlgo: string,
                         _database: string, _fdrLevel: number): Promise<number | null>;
}

