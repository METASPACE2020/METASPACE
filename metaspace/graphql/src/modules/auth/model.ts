import 'reflect-metadata';
import {Column, Entity, PrimaryColumn} from 'typeorm';
import {Moment} from 'moment';
import {MomentValueTransformer} from '../../utils/MomentValueTransformer';

@Entity()
export class Credentials {

  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', nullable: true })
  hash: string | null;

  @Column({ type: 'text', name: 'google_id', nullable: true })
  googleId: string | null;

  @Column({ type: 'text', name: 'email_verification_token', nullable: true })
  emailVerificationToken: string | null;

  @Column({ type: 'timestamp without time zone', name: 'email_verification_token_expires',
    nullable: true, transformer: new MomentValueTransformer() })
  emailVerificationTokenExpires?: Moment | null;

  @Column({ type: 'boolean', name: 'email_verified', default: false })
  emailVerified: boolean;

  @Column({ type: 'text', name: 'reset_password_token', nullable: true })
  resetPasswordToken: string | null;

  @Column({ type: 'timestamp without time zone', name: 'reset_password_token_expires',
    nullable: true, transformer: new MomentValueTransformer() })
  resetPasswordTokenExpires: Moment | null;
}
