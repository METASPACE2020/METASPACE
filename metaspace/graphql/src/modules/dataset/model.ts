import {Column, Entity, JoinColumn, JoinTable, ManyToOne, OneToMany, PrimaryColumn} from 'typeorm';
import {Group} from '../group/model';
import {User} from '../user/model';
import {Project} from '../project/model';

@Entity()
export class Dataset {

  @PrimaryColumn({ type: 'text' })
  id: string;

  @Column({ type: 'uuid', name: 'user_id' })
  userId: string; // dataset submitter and owner -> edit rights

  @ManyToOne(type => User, user => user.datasets)
  @JoinColumn({ name: 'user_id' })
  user: User;

  @Column({ type: 'uuid', name: 'group_id', nullable: true })
  groupId: string | null; // dataset belongs to group -> all members have view rights

  @ManyToOne(type => Group)
  @JoinColumn({ name: 'group_id' })
  group: Group;

  @Column({ default: false, name: 'group_approved' })
  groupApproved: boolean; // true when submitter is a group member

  // when user manually inputs PI details
  @Column({ type: 'text', name: 'pi_name', nullable: true })
  piName: string | null;

  @Column({ type: 'text', name: 'pi_email', nullable: true })
  piEmail: string | null;

  @OneToMany(type => DatasetProject, datasetProject => datasetProject.dataset)
  @JoinTable({ name: 'dataset_project' })
  datasetProjects: DatasetProject[];
}

@Entity({ name: 'dataset_project' })
export class DatasetProject {

  @PrimaryColumn({ type: 'text', name: 'dataset_id' })
  datasetId: string;

  @ManyToOne(type => Dataset)
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @PrimaryColumn({ type: 'uuid', name: 'project_id' })
  projectId: string;

  @ManyToOne(type => Project)
  @JoinColumn({ name: 'project_id' })
  project: Project;

  @Column({ type: 'boolean' })
  approved: Boolean;
}

export type DatasetStatus = 'QUEUED' | 'ANNOTATING' | 'FINISHED' | 'FAILED';
// This isn't an Entity because datasets in the "public" schema are owned by sm-engine and aren't managed by TypeORM
export interface EngineDataset {
  id: string;
  name: string | null;
  input_path: string | null;
  metadata: object | null;
  config: object | null;
  upload_dt: Date | null;
  status: DatasetStatus | null;
  optical_image: string | null;
  transform: number[] | null;
  is_public: boolean;
  acq_geometry: object | null;
  ion_img_storage_type: string;
  thumbnail: string | null;
  ion_thumbnail: string | null;
  mol_dbs: string[];
  adducts: string[];
}

export interface EngineOpticalImage {
  id: string;
  ds_id: string;
  type: string;
  zoom: number;
  scale: number;
  width: number;
  height: number;
  transform: number[][];
}
