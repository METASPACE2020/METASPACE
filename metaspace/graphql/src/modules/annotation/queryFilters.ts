import * as _ from 'lodash';
import {Context} from '../../context';
import {ArgsFromBinding} from '../../bindingTypes';
import {Query} from '../../binding';
import {ColocAnnotation, ColocJob, Ion} from './model';
import {UserError} from 'graphql-errors';
import config from '../../utils/config';
import {ESAnnotation} from '../../../esConnector';

type Args = (ArgsFromBinding<Query['allAnnotations']>
          | ArgsFromBinding<Query['countAnnotations']>);
interface FilterResult {
  args: Args;
  postprocess?(annotations: ESAnnotation[]): ESAnnotationWithColoc[] ;
}

export interface ESAnnotationWithColoc extends ESAnnotation {
  _cachedColocCoeff: number | null;
  _isColocReference: boolean;
  getColocalizationCoeff(_colocalizedWith: string, _colocalizationAlgo: string,
                         _database: string, _fdrLevel: number): Promise<number | null>;
}

const setOrMerge = <T>(obj: any, path: string, value: T, onConflict = (val: T, oldVal: T) => val) => {
  let newVal = value;
  if (_.hasIn(obj, path)) {
    const oldVal = _.get(obj, path);
    newVal = onConflict(value, oldVal);
  }
  return _.merge({}, obj, _.set({}, path, newVal));
};

interface AnnotationAndIons {
  annotation: ColocAnnotation;
  ionsById: Map<number, Ion>;
  lookupIon: (formula: string, chemMod: string, neutralLoss: string, adduct: string) => Ion | undefined;
}

const getColocAnnotation = async (context: Context, datasetId: string, fdrLevel: number, database: string | null,
                           colocalizedWith: string, colocalizationAlgo: string): Promise<AnnotationAndIons | null> => {
  const args = [datasetId, fdrLevel, database, colocalizedWith, colocalizationAlgo] as const;
  return await context.contextCacheGet('getColocAnnotation', args,
    async (datasetId, fdrLevel, database, colocalizedWith, colocalizationAlgo) => {
      const annotation = await context.entityManager.createQueryBuilder(ColocAnnotation, 'colocAnnotation')
        .innerJoinAndSelect('colocAnnotation.colocJob', 'colocJob')
        .innerJoin(Ion, 'ion', 'colocAnnotation.ionId = ion.id')
        .where('colocJob.datasetId = :datasetId AND colocJob.fdr = :fdrLevel AND colocJob.molDb = :database ' +
          'AND colocJob.algorithm = :colocalizationAlgo AND ion.ion = :colocalizedWith',
          { datasetId, fdrLevel, database, colocalizationAlgo, colocalizedWith })
        .getOne();
      if (annotation != null) {
        const ions = await context.entityManager.findByIds(Ion, [annotation.ionId, ...annotation.colocIonIds]);
        const ionsByFormula = _.groupBy(ions, 'formula');
        return {
          annotation,
          ionsById: new Map<number, Ion>(ions.map(ion => [ion.id, ion] as [number, Ion])),
          lookupIon: (formula: string, chemMod: string, neutralLoss: string, adduct: string) => {
            return (ionsByFormula[formula] || [])
              .find(ion => ion.chemMod === chemMod && ion.neutralLoss === neutralLoss && ion.adduct === adduct);
          },
        };
      } else {
        return null;
      }
    });
};

const getColocSampleIons = async (context: Context, datasetId: string, fdrLevel: number, database: string | null,
                           colocalizationAlgo: string) => {
  const result = await context.entityManager.findOne(ColocJob,
    {datasetId, fdr: fdrLevel, molDb: database, algorithm: colocalizationAlgo},
    { select: ['sampleIonIds'] });
  if (result == null) {
    return null;
  } else {
    const ions = await context.entityManager.findByIds(Ion, result.sampleIonIds, {select: ['ion']});
    return ions.map(({ion}) => ion);
  }
};

const getColocCoeffInner = (baseAnnotation: ColocAnnotation, otherIonId: number) => {
  const { ionId, colocIonIds, colocCoeffs } = baseAnnotation;

  if (otherIonId === ionId) {
    return 1; // Same molecule
  } else {
    const coeff_idx = colocIonIds.indexOf(otherIonId);
    if (coeff_idx !== -1) {
      return colocCoeffs[coeff_idx];
    } else {
      return null;
    }
  }
};


/**
 * Apply filters based on data that's not available in ElasticSearch by finding the matching data in postgres,
 * then using it to add additional filters to the ElasticSearch query. Some filters may also require that extra data
 * is added to the found annotations for other resolvers to use.
 *
 * When the datasetId is provided, ES can easily handle large numbers of arguments in "terms" queries.
 * Querying with 28000 valid values of sfAdduct ran in 55ms. It would cost a significant amount of disk space to
 * index colocalized molecules, so filtering only in postgres seems to be the better option.
 * */
export const applyQueryFilters = async (context: Context, args: Args): Promise<FilterResult> => {

  const datasetId = args.datasetFilter && args.datasetFilter.ids;
  const {
    fdrLevel = 0.1,
    database = null,
    colocalizedWith = null,
    colocalizationSamples = false,
  } = args.filter || {};
  const colocalizationAlgo =
    args.filter && args.filter.colocalizationAlgo
    || config.metadataLookups.defaultColocalizationAlgo;
  let newArgs = args;
  let postprocess: FilterResult['postprocess'];


  if (datasetId != null && colocalizationAlgo != null && colocalizationSamples) {
    const samples = await getColocSampleIons(context, datasetId, fdrLevel, database, colocalizationAlgo);
    if (samples != null) {
      newArgs = setOrMerge(newArgs, 'filter.ion', samples, _.intersection);
    } else {
      newArgs = setOrMerge(newArgs, 'filter.ion', []);
    }
  }

  if (datasetId != null && colocalizationAlgo != null && colocalizedWith != null) {
    const annotationAndIons = await getColocAnnotation(context, datasetId, fdrLevel, database, colocalizedWith, colocalizationAlgo);

    if (annotationAndIons != null) {
      const {annotation, ionsById, lookupIon} = annotationAndIons;
      const {offset, limit} = args as ArgsFromBinding<Query['allAnnotations']>;
      const colocIons = _.uniq([annotation.ionId, ...annotation.colocIonIds])
        .map(ionId => {
          const ion = ionsById.get(ionId);
          return ion != null ? ion.ion : null
        })
        .filter(ion => ion != null);

      newArgs = setOrMerge(newArgs, 'filter.ion', colocIons, _.intersection);
      // Always select 1000 annotations so that sorting by colocalizationCoeff doesn't just sort per-page
      newArgs = setOrMerge(newArgs, 'offset', 0);
      newArgs = setOrMerge(newArgs, 'limit', 1000);

      postprocess = (annotations: ESAnnotation[]): ESAnnotationWithColoc[] => {
        let newAnnotations: ESAnnotationWithColoc[] = annotations.map(ann => {
          const ion = lookupIon(ann._source.formula, ann._source.chem_mod, ann._source.neutral_loss, ann._source.adduct);
          return {
            ...ann,
            _cachedColocCoeff: ion != null ? getColocCoeffInner(annotation, ion.id) : null,
            _isColocReference: ion != null && ion.id === annotation.ionId,
            async getColocalizationCoeff(_colocalizedWith: string, _colocalizationAlgo: string, _database: string, _fdrLevel: number) {
              if (_colocalizedWith === colocalizedWith && _colocalizationAlgo === colocalizationAlgo && _database === database && _fdrLevel === fdrLevel) {
                return this._cachedColocCoeff;
              } else {
                throw new UserError('colocalizationCoeff arguments must match the parent allAnnotations filter values');
                // return getColocCoeff(context, datasetId, _colocalizedWith, ann._source.sf_adduct, _colocalizationAlgo, _database, _fdrLevel)
              }
            },
          }
        });

        if (!('orderBy' in args) || args.orderBy === 'ORDER_BY_COLOCALIZATION') {
          const order = 'sortingOrder' in args && args.sortingOrder === 'ASCENDING' ? 1 : -1;

          newAnnotations = _.sortBy(newAnnotations, ann => {
            if (ann._isColocReference) {
              // Always show reference annotations as the most colocalized,
              // even if there are other annotations with a 1.0 colocalizationCoeff
              return Infinity * order;
            } else {
              return (ann._cachedColocCoeff != null ? ann._cachedColocCoeff : -1) * order;
            }
          })
        }
        if (offset) {
          newAnnotations = newAnnotations.slice(offset);
        }
        if (limit) {
          newAnnotations = newAnnotations.slice(0, limit);
        }

        return newAnnotations;
      }
    } else {
      newArgs = setOrMerge(newArgs, 'filter.ion', []);
    }
  }

  return {args: newArgs, postprocess};
};
