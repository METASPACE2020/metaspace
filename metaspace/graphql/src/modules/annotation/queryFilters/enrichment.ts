import { Context } from '../../../context'
import { QueryFilterArgs, QueryFilterResult } from './types'
import {
  EnrichmentBootstrap,
  EnrichmentDBMoleculeMapping as EnrichmentDBMoleculeMappingModel,
} from '../../enrichmentdb/model'
import { uniq } from 'lodash'
import { setOrMerge } from '../../../utils/setOrMerge'
import { EngineDataset } from '../../engine/model'

export const applyEnrichmentTermFilter =
    async(context: Context, args: QueryFilterArgs): Promise<QueryFilterResult> => {
      const termId = args.filter && args.filter.termId
      const databaseId = args.filter && args.filter.databaseId

      if (termId) {
        let adducts: any = []

        const enrichmentTermsMapping = await context.entityManager.createQueryBuilder(EnrichmentDBMoleculeMappingModel,
          'mapping')
          .leftJoin('mapping.enrichmentTerm', 'terms')
          .select(['mapping.formula', 'mapping.id'])
          .distinct(true)
          .where('mapping.enrichmentTermId = :termId', { termId: termId })
          .getRawMany()
        const formulas = enrichmentTermsMapping.map((term: any) => term.mapping_formula)
        const ids : any = args?.datasetFilter?.ids

        // restrict to used adducts if dataset id filter passed (based on bootstrapping)
        if (ids && databaseId) {
          const bootstrap = await context.entityManager
            .createQueryBuilder(EnrichmentBootstrap,
              'bootstrap')
            .leftJoin('bootstrap.enrichmentDBMoleculeMapping', 'enrichmentDBMoleculeMapping')
            .select(['bootstrap.formulaAdduct', 'bootstrap.datasetId'])
            .distinct(true)
            .where((qb : any) => {
              qb.where('bootstrap.datasetId  IN (:...ids)', { ids: ids.split('|') })
                .andWhere('bootstrap.enrichmentDbMoleculeMappingId  IN (:...termIds)',
                  { termIds: enrichmentTermsMapping.map((term: any) => term.mapping_id) })
                .andWhere('enrichmentDBMoleculeMapping.molecularDbId  = :dbId',
                  { dbId: databaseId })
            })
            .getRawMany()

          // get dataset polarity
          const datasets = await context.entityManager
            .createQueryBuilder(EngineDataset,
              'engineDataset')
            .select(['engineDataset.id', 'engineDataset.metadata'])
            .where((qb : any) => {
              qb.where('engineDataset.id  IN (:...ids)', { ids: ids.split('|') })
            })
            .getRawMany()
          const datasetHash : any = {}
          datasets.forEach((dataset: any) => {
            datasetHash[dataset.engineDataset_id] = dataset.engineDataset_metadata?.MS_Analysis?.Polarity
            datasetHash[dataset.engineDataset_id] =
                  datasetHash[dataset.engineDataset_id] === 'Positive' ? '+' : '-'
          })

          bootstrap.forEach((item: any) => {
            const auxAdduct : string = item.bootstrap_formula_adduct
            const dsId : string = item.bootstrap_dataset_id
            formulas.forEach((formula: any) => {
              if (auxAdduct.indexOf(formula) !== -1) {
                adducts.push(`${auxAdduct}${datasetHash[dsId]}`) // add polarity
              }
            })
            return auxAdduct
          })
          adducts = uniq(adducts)
          return { args: setOrMerge(args, 'filter.ion', adducts) }
        }

        return { args: setOrMerge(args, 'filter.sumFormula', formulas) }
      }
      return { args }
    }
