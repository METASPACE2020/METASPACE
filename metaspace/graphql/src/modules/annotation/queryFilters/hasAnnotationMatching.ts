import {Context} from '../../../context';
import {QueryFilterArgs, QueryFilterResult} from './types';
import {esCountMatchingAnnotationsPerDataset} from '../../../../esConnector';
import * as _ from 'lodash';


export const applyHasAnnotationMatchingFilter =
  async (context: Context, args: QueryFilterArgs): Promise<QueryFilterResult> => {
    const {datasetFilter, ...otherArgs} = args;

    if (datasetFilter != null && datasetFilter.hasAnnotationMatching != null) {
      const {hasAnnotationMatching, ...otherDatasetFilters} = datasetFilter;
      const result = await esCountMatchingAnnotationsPerDataset({
        datasetFilter: otherDatasetFilters,
        // simpleQuery: args.simpleQuery,
        filter: hasAnnotationMatching,
      }, context.user);

      const datasetIds = Object.keys(result);
      const ids = otherDatasetFilters.ids
        ? _.intersection(datasetIds, otherDatasetFilters.ids.split('|')).join('|')
        : datasetIds.join('|');

      return {
        args: {
          ...otherArgs,
          datasetFilter: {
            ...otherDatasetFilters,
            ids,
          }
        },
      };
    } else {
      return { args };
    }
  };
