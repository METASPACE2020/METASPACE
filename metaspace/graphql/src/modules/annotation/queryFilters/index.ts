import {Context} from '../../../context';
import {QueryFilterArgs, QueryFilterResult, PostProcessFunc} from './types';
import {applyColocalizationSamplesFilter} from './colocalizationSamples';
import {applyColocalizedWithFilter} from './colocalizedWith';
import * as _ from 'lodash';

export {ESAnnotationWithColoc} from './types';

const queryFilters = [
  applyColocalizationSamplesFilter,
  applyColocalizedWithFilter,
];

/**
 * Augment filters based on data that's not easily queryable in ElasticSearch, e.g. data that is only available in
 * postgres, or data that requires multiple queries in ElasticSearch. Some filters may also require that extra data
 * is added to the found annotations for other resolvers to use.
 *
 * When the datasetId is provided, ES can easily handle large numbers of arguments in "terms" queries.
 * Querying with 28000 valid values of sfAdduct ran in 55ms. It would cost a significant amount of disk space to
 * index colocalized molecules, so filtering only in postgres seems to be the better option.
 * */
export const applyQueryFilters = async (context: Context, args: QueryFilterArgs): Promise<QueryFilterResult> => {
  let newArgs = args;
  let postprocessFuncs: PostProcessFunc[] = [];

  for (const filter of queryFilters) {
    const result = await filter(context, newArgs);
    if (result.args != null) {
      newArgs = result.args;
    }
    if (result.postprocess != null) {
      postprocessFuncs.push(result.postprocess);
    }
  }

  return { args: newArgs, postprocess: _.flow(postprocessFuncs) };
};
