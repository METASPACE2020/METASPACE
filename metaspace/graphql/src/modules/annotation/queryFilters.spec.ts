import * as _ from 'lodash';
import {
  anonContext,
  onAfterAll,
  onAfterEach,
  onBeforeAll,
  onBeforeEach,
  testEntityManager,
} from '../../tests/graphqlTestEnvironment';
import {ColocAnnotation, ColocJob, Ion} from './model';
import {ArgsFromBinding} from '../../bindingTypes';
import {Query} from '../../binding';
import {applyQueryFilters} from './queryFilters';
import {ESAnnotation} from '../../../esConnector';
import {DeepPartial} from 'typeorm';

type Args = ArgsFromBinding<Query['allAnnotations']>
  | ArgsFromBinding<Query['countAnnotations']>;

describe('annotation/queryFilters applyQueryFilters', () => {
  beforeAll(onBeforeAll);
  afterAll(onAfterAll);
  // beforeEach(onBeforeEach);
  afterEach(onAfterEach);

  let ions: Ion[];
  let job: ColocJob;

  beforeEach(async () => {
    await onBeforeEach();

    ions = testEntityManager.create(Ion, [1,2,3,4].map(i => ({
      ion: `H${i}O+H+`,
      formula: `H${i}O`,
      adduct: '+H',
      charge: 1
    })));
    await testEntityManager.insert(Ion, ions);

    job = testEntityManager.create(ColocJob, {
      datasetId: 'datasetId',
      molDb: 'HMDB-v4',
      fdr: 0.1,
      algorithm: 'cosine',
      sampleIonIds: [ions[1].id, ions[2].id],
    });
    const res = await testEntityManager.insert(ColocJob, job);

    const annotations = testEntityManager.create(ColocAnnotation, _.range(ions.length).map(i => ({
      colocJobId: job.id,
      ionId: ions[i].id,
      colocIonIds: [...ions, ...ions].slice(i+1, i+3).map(ion => ion.id),
      colocCoeffs: [0.9, 0.7],
    })));
    await testEntityManager.insert(ColocAnnotation, annotations);
  });

  it('should transform the args of a request with a colocalizedWith filter', async () => {
    const argsWithColocWith: Args = {
      datasetFilter: { ids: job.datasetId },
      filter: {
        database: job.molDb!,
        fdrLevel: job.fdr,
        colocalizedWith: ions[1].ion,
        colocalizationAlgo: job.algorithm,
      }
    };

    const {args, postprocess} = await applyQueryFilters(anonContext, argsWithColocWith);

    expect(args).toMatchSnapshot();
    expect(args).toMatchObject({
      offset: 0,
      limit: 1000,
      filter: {
        ion: ions.slice(1,4).map(ion => ion.ion),
      }
    });
  });

  it('should transform the args of a request with a colocalizationSamples filter', async () => {
    const argsWithColocWith: Args = {
      datasetFilter: { ids: job.datasetId },
      filter: {
        database: job.molDb!,
        fdrLevel: job.fdr,
        colocalizationSamples: true,
        colocalizationAlgo: job.algorithm,
      }
    };

    const {args, postprocess} = await applyQueryFilters(anonContext, argsWithColocWith);

    expect(args).toMatchSnapshot();
    expect(args).toMatchObject({
      filter: {
        ion: ions.slice(1,3).map(ion => ion.ion)
      }
    });
  });

  it('should add colocalizationCoeffs to annotations', async () => {
    const argsWithColocWith: Args = {
      datasetFilter: { ids: job.datasetId },
      filter: {
        database: job.molDb!,
        fdrLevel: job.fdr,
        colocalizedWith: ions[1].ion,
        colocalizationAlgo: job.algorithm,
      }
    };
    const {args, postprocess} = await applyQueryFilters(anonContext, argsWithColocWith);
    const annotations: DeepPartial<ESAnnotation>[] = [
      {_source: {ion: ions[2].ion}},
      {_source: {ion: ions[3].ion}},
      {_source: {ion: ions[1].ion}},
    ];

    const result = postprocess!(annotations as ESAnnotation[]);

    expect(result).toMatchObject([
      {_source: {ion: ions[1].ion}, _isColocReference: true},
      {_source: {ion: ions[2].ion}, _isColocReference: false},
      {_source: {ion: ions[3].ion}, _isColocReference: false},
    ]);
    expect(await result[0].getColocalizationCoeff(ions[1].ion, job.algorithm, job.molDb!, job.fdr)).toEqual(1);
    expect(await result[1].getColocalizationCoeff(ions[1].ion, job.algorithm, job.molDb!, job.fdr)).toEqual(0.9);
    expect(await result[2].getColocalizationCoeff(ions[1].ion, job.algorithm, job.molDb!, job.fdr)).toEqual(0.7);
  });
});
