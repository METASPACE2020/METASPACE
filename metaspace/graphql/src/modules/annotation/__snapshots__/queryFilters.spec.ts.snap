// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotation/queryFilters applyQueryFilters should transform the args of a request with a colocalizationSamples filter 1`] = `
Object {
  "datasetFilter": Object {
    "ids": "datasetId",
  },
  "filter": Object {
    "colocalizationAlgo": "cosine",
    "colocalizationSamples": true,
    "database": "HMDB-v4",
    "fdrLevel": 0.1,
    "ion": Array [
      "H2O+H+",
      "H3O+H+",
    ],
  },
}
`;

exports[`annotation/queryFilters applyQueryFilters should transform the args of a request with a colocalizedWith filter 1`] = `
Object {
  "datasetFilter": Object {
    "ids": "datasetId",
  },
  "filter": Object {
    "colocalizationAlgo": "cosine",
    "colocalizedWith": "H2O+H+",
    "database": "HMDB-v4",
    "fdrLevel": 0.1,
    "ion": Array [
      "H2O+H+",
      "H3O+H+",
      "H4O+H+",
    ],
  },
  "limit": 1000,
  "offset": 0,
}
`;
